src/shadowbox/server/manager_service.ts

```diff
@@ -125,6 +128,19 @@ function validateAccessKeyId(accessKeyId: unknown): string {
   return accessKeyId;
 }
 
+function validateDataLimit(limit: unknown): DataLimit {
```

Comment:
You could use a [type-guard](https://basarat.gitbook.io/typescript/type-system/typeguard#user-defined-type-guards) avoid reasigning the returned value:

```typescript
function isDataLimit(limit: unknown): is DataLimit {
```

Reply:
I didn't use a type guard because the function would always either throw or return true.  I use exceptions here since this function knows what the exact error this is, allowing us to return consistent errors on ever endpoint which takes a user-supplied data limit. 

Using a type guard would set up a situation where we have

```
if isDataLimit(req.params.limit) {
  // The entire rest of the function which uses the limit
}
// some other code which looks like it will run if the limit isn't valid, but actually won't ever be run
// unless isDataLimit is true because otherwise we'll throw and enter the exception handler
```

This looks really weird at the call-site to me.

Reply:
I didn't use a type guard because the function would always either throw or return true.  I use exceptions here since this function knows what the exact error this is, allowing us to return consistent errors on ever endpoint which takes a user-supplied data limit. 

Using a type guard would set up a situation where we have

```
if isDataLimit(req.params.limit) {
  // The entire rest of the function which uses the limit
}
// some other code which looks like it will run if the limit isn't valid, but actually won't ever be run
// unless isDataLimit is true because otherwise we'll throw and enter the exception handler
```

This looks really weird at the call-site to me.

Reply:
Fair enough, I missed that you were throwing specific errors.